### PR TITLE
120 implement unit tests for task  #108 field validation

### PR DESCRIPTION
**Pull request by Artem3**
Please reply on questions in  comments in the code and I'll refactor.
Don't hesitate contact me  :)

### CODE REVIEW

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java [hunk 1]
```diff
@@ -101,8 +134,56 @@ public void When_AddingNewTask_Expect_NewTaskAddedToDB() throws InterruptedExcep
...
+    @Test
+    public void When_AddingNewTask_and_Name_Already_Exists_Expect_Warning() {
+        // I have to emulate existing name on DB. How ?! I need smth  like
+        // TaskTemplateRepository.isNameOndb(name);
+    }
+
     private void closeKeyboard() throws InterruptedException {
         closeSoftKeyboard();
-        Thread.sleep(1000);
+        Thread.sleep(1000); // I read that Espress thread is syncronized with main flow
```
Comment (malsolec) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java]: I think it was added for Travis, normal sleep was not enough there. You could delete it in this test and check if Travis is passing, anyway. If not, you will have the answer.

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java [hunk 2]
```diff
@@ -101,8 +134,56 @@ public void When_AddingNewTask_Expect_NewTaskAddedToDB() throws InterruptedExcep
...
+                .perform(click());
+
+        onView(withId(R.id.id_et_task_name)).check
+                (matches(hasErrorText(
+                        activityRule.getActivity().getString(R.string.only_letters_msg))));
+    }
+
+    @Test
+    public void When_AddingNewTask_and_Name_Already_Exists_Expect_Warning() {
+        // I have to emulate existing name on DB. How ?! I need smth  like
```
Comment (malsolec) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java]: Yes, right, This test should be added with task, #119. So you do no need this now.

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java [hunk 3]
```diff
@@ -12,33 +14,64 @@
...
     public ActivityTestRule<TaskCreateActivity> activityRule = new ActivityTestRule<>(
             TaskCreateActivity.class, true, true);
-
-    private static final String EXPECTED_NAME = "TEST TASK";
-
     private TaskTemplateRepository taskTemplateRepository;
     private Long idToDelete;
 
+    // How about @inject this custom matcher through annotation ? No ? just usual extract to class ?
+    private static Matcher<View> hasErrorText(final String expectedErrorText) {
```
Comment (malsolec) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java]: If it is possible, go on. Sounds good.

Comment (Artem3) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java]: Extracted to separate package and class.

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java (hunk 2 above)
Comment (Artem3) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java]: Removed.

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java (hunk 1 above)
Comment (Artem3) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/TaskCreateActivityTest.java]: Method removed, Travis is OK about that :)

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                CharSequence error = ((EditText) view).getError();
+                if (error == null) {
+                    return false;
+                }
+
+                String actualError = error.toString();
+                return expectedErrorText.equals(actualError);
+            }
+            @Override
+            public void describeTo(Description description) { }
```
Comment (malsolec) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java]: Codacy require method no to be empty and in that case it is true. Please update this method with you custom message for matcher. Description is here: http://www.planetgeek.ch/2012/03/07/create-your-own-matcher/

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import org.hamcrest.TypeSafeMatcher;
+
+/**
+ * Use this custom matcher when you need match/assert <b>error text</b> on mandatory(*) EditText
+ * fields in xml layout.
+ */
+
+public class ErrorTextMatcher {
+
+    public static Matcher<View> hasErrorText(final String expectedErrorText) {
```
Comment (malsolec) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java]: Codacy also wants to add private constructor for this method to ensure it cannot be initialise. We have already followed this tip in other cases so we could do this this time.

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+                    return false;
+                }
+
+                String actualError = error.toString();
+                return expectedErrorText.equals(actualError);
+            }
+            @Override
+            public void describeTo(Description description) { }
+        };
+    }
```
Comment (malsolec) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java]: Codacy also wants to add private constructor for this method to ensure it cannot be initialise. We have already followed this tip in other cases so we could do this this time.

Add at the bottom empty private constructor.

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java (hunk 4 above)
Comment (Artem3) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java]: Message added.

File: Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java (hunk 6 above)
Comment (Artem3) [inline Friendly-plans/app/src/androidTest/java/pg/autyzm/friendly_plans/matchers/ErrorTextMatcher.java]: Private constructor added.

**Pull request merged**